Sources/NIO/EventLoopFuture.swift

```diff
@@ -1044,7 +1054,83 @@ extension EventLoopFuture {
     ///     - on: The `EventLoop` on which the new `EventLoopFuture` callbacks will fire.
     /// - Returns: A new `EventLoopFuture` with all of the values fulfilled by the provided futures.
     public static func whenAllSucceed(_ futures: [EventLoopFuture<Value>], on eventLoop: EventLoop) -> EventLoopFuture<[Value]> {
-        return .reduce(into: [], futures, on: eventLoop) { (results, value) in results.append(value) }
```

Comment:
Both of these methods were using `.reduce`, and I can't help but wonder if perhaps the performance boost could have been done in that method?

Reply:
I'm sure `reduce` and `fold` have potential for further optimization, but I don't think their performance will come anywhere close to this bespoke implementation. For example, the argument to `fold` is a closure that returns a future. I.e. we'd need to create at least one extra future (several more with the current implementation of `fold`, actually) for each element in the array. That alone generates so much overhead that it can't match the performance of bespoke implementations.

There could be a point made about when the performance of a bespoke implementation is needed, however, Vapor & Co. process a ton of already-fulfilled futures where this is significant.